File: frontend/src/lib/permissions.ts
import type {
  HomeCardKey,
  PermAction,
  PermResource,
  PermissionMap,
  Profile,
  Role,
} from "@/types";

// ─── Catalog (mirrors backend app/core/permissions.py) ─────────────────────────
// Single source of truth for which actions exist per resource on the frontend.
// The settings matrix is rendered from the live backend catalog, but this local
// copy drives tab/button gating without an extra round-trip.

export const PERMISSION_CATALOG: Record<PermResource, PermAction[]> = {
  home: ["view"],
  metrics: ["view", "create", "edit", "delete"],
  tasks: ["view", "create", "edit", "delete"],
  finance: ["view", "create", "edit", "delete"],
  clients: ["view", "create", "edit", "delete"],
};

// Resources every authenticated user can always view, regardless of role/map.
const ALWAYS_VIEWABLE: PermResource[] = ["home"];

export const RESOURCE_LABELS: Record<PermResource, string> = {
  home: "Home",
  metrics: "Métricas",
  tasks: "Tarefas",
  finance: "Finanças",
  clients: "Clientes",
};

export const ACTION_LABELS: Record<PermAction, string> = {
  view: "Ver",
  create: "Criar",
  edit: "Editar",
  delete: "Excluir",
};

// ─── Access resolution ─────────────────────────────────────────────────────────

export function isManager(role: Role | undefined | null): boolean {
  return role === "manager";
}

/** Resolve whether a profile may perform `action` on `resource`. */
export function can(
  profile: Pick<Profile, "role" | "permissions"> | null | undefined,
  resource: PermResource,
  action: PermAction = "view"
): boolean {
  if (!profile) return false;
  if (profile.role === "manager") return true;
  if (action === "view" && ALWAYS_VIEWABLE.includes(resource)) return true;
  const entry = profile.permissions?.[resource];
  return Boolean(entry?.[action]);
}

// ─── Home cards ────────────────────────────────────────────────────────────────

export const HOME_CARDS: HomeCardKey[] = ["goals", "tasks", "finance"];

export const HOME_CARD_LABELS: Record<HomeCardKey, string> = {
  goals: "Metas Ativas",
  tasks: "Tarefas Prioritárias",
  finance: "Receita do Mês",
};

/** Resolve whether a profile may see a given Home card (managers see all). */
export function canSeeHomeCard(
  profile: Pick<Profile, "role" | "home_cards"> | null | undefined,
  card: HomeCardKey
): boolean {
  if (!profile) return false;
  if (profile.role === "manager") return true;
  return profile.home_cards?.[card] !== false;
}

/** Build an empty permission map (every action false) for the editable matrix. */
export function emptyPermissionMap(): PermissionMap {
  const map: PermissionMap = {};
  for (const [resource, actions] of Object.entries(PERMISSION_CATALOG)) {
    if (ALWAYS_VIEWABLE.includes(resource as PermResource)) continue;
    map[resource as PermResource] = Object.fromEntries(
      actions.map((a) => [a, false])
    );
  }
  return map;
}

File: backend/app/core/permissions.py
"""Role & permission model — single source of truth for access control.

Two roles exist:

- ``manager``      → full access to every resource and action; the JSONB
                     permission map is ignored entirely.
- ``collaborator`` → access is driven by a per-resource, per-action map.

The frontend mirrors this catalog (TypeScript) and also fetches it live via
``GET /api/v1/admin/catalog`` so the settings matrix renders dynamically.
"""

from __future__ import annotations

from enum import Enum


class Role(str, Enum):
    MANAGER = "manager"
    COLLABORATOR = "collaborator"


class Action(str, Enum):
    VIEW = "view"
    CREATE = "create"
    EDIT = "edit"
    DELETE = "delete"


class Resource(str, Enum):
    HOME = "home"
    METRICS = "metrics"
    TASKS = "tasks"
    FINANCE = "finance"
    CLIENTS = "clients"


# Which actions are meaningful for each resource. ``home`` is view-only and is
# always granted to any authenticated user.
PERMISSION_CATALOG: dict[Resource, list[Action]] = {
    Resource.HOME: [Action.VIEW],
    Resource.METRICS: [Action.VIEW, Action.CREATE, Action.EDIT, Action.DELETE],
    Resource.TASKS: [Action.VIEW, Action.CREATE, Action.EDIT, Action.DELETE],
    Resource.FINANCE: [Action.VIEW, Action.CREATE, Action.EDIT, Action.DELETE],
    Resource.CLIENTS: [Action.VIEW, Action.CREATE, Action.EDIT, Action.DELETE],
}

# Resources that every authenticated user can always view, regardless of role
# or permission map.
_ALWAYS_VIEWABLE: frozenset[Resource] = frozenset({Resource.HOME})


def catalog_payload() -> list[dict]:
    """Serialize the catalog for the frontend (stable, ordered)."""
    return [
        {"resource": resource.value, "actions": [a.value for a in actions]}
        for resource, actions in PERMISSION_CATALOG.items()
    ]


def normalize(raw: object) -> dict[str, dict[str, bool]]:
    """Sanitize an incoming permission map against the catalog.

    Drops unknown resources/actions and coerces values to bool, so malformed
    or malicious payloads can never widen access beyond the known catalog.
    """
    if not isinstance(raw, dict):
        return {}

    clean: dict[str, dict[str, bool]] = {}
    for resource, actions in PERMISSION_CATALOG.items():
        if resource in _ALWAYS_VIEWABLE:
            continue
        entry = raw.get(resource.value)
        if not isinstance(entry, dict):
            continue
        resource_perms: dict[str, bool] = {}
        for action in actions:
            resource_perms[action.value] = bool(entry.get(action.value, False))
        clean[resource.value] = resource_perms
    return clean


class UserPermissions:
    """Resolves access questions for a single user."""

    def __init__(self, role: str, permissions: object) -> None:
        self.role = role
        self._perms = permissions if isinstance(permissions, dict) else {}

    @property
    def is_manager(self) -> bool:
        return self.role == Role.MANAGER.value

    def can(self, resource: Resource, action: Action) -> bool:
        if self.is_manager:
            return True
        if action == Action.VIEW and resource in _ALWAYS_VIEWABLE:
            return True
        entry = self._perms.get(resource.value)
        if not isinstance(entry, dict):
            return False
        return bool(entry.get(action.value, False))

    def accessible_tabs(self) -> list[str]:
        """Resources the user can at least view."""
        return [
            resource.value
            for resource in PERMISSION_CATALOG
            if self.can(resource, Action.VIEW)
        ]
